src/Translator.php

```diff
@@ -0,0 +1,152 @@
+     *
+     * @var string
+     */
+    private $fallback;
+
+    /**
+     * Path where all translation are stored
+     * Can be null because translation can be add at runtime.
+     *
+     * @var string|null
+     */
+    private $translation_path;
+
+    /**
+     * Format for ConfigTrait to read translations.
+     *
+     * @var string
+     */
+    private $translation_format;
+
+    public function getLanguage(): string
+    {
+        return $this->language;
+    }
+
+    public function getFallback(): string
+    {
+        return $this->fallback;
+    }
+
+    /**
+     * Translator constructor.
+     *
+     * @param string|null $translation_path root path of translations
+     * @param string      $format           format for ConfigTrait loader
+     */
+    public function __construct(string $translation_path = null, string $format = 'php-inline')
+    {
+        $this->translation_path = $translation_path;
+        $this->translation_format = $format;
+    }
+
+    /**
+     * {@inheritdoc}
+     * @throws Exception
+     */
+    public function set(string $ISOCode, string $fallbackISOCode = null)
+    {
+        $this->language = $ISOCode;
+
+        $this->fallback = $fallbackISOCode ?? $this->language;
+
+        // if no base path is specified don't load
+        if (!$this->translation_path) {
+            return;
+        }
+
+        $ext = 'php';
+
+        switch ($this->translation_format) {
+            case 'php':
+            case 'php-inline':
+                //$ext = 'php';
+                break;
+
+            case 'json':
+                $ext = 'json';
+                break;
+
+            case 'yaml':
+                $ext = 'yaml';
+                break;
+        }
+
+        $language_files = [
+            $this->translation_path.DIRECTORY_SEPARATOR.$this->language.'.'.$ext,
+            $this->translation_path.DIRECTORY_SEPARATOR.$this->fallback.'.'.$ext,
+        ];
+
+        $language_files = array_unique($language_files);
+
+        $this->readConfig($language_files, $this->translation_format);
+
+        $this->translation = $this->config;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addOne(string $string, array $translations, string $domain = 'atk4')
+    {
+        if (array_key_exists($string, $this->translation)) {
+            throw new Exception('Translation already exists');
+        }
+
+        $this->translation[$domain][$string] = $translations;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function translate(string $message, int $count = 1, ?string $context = NULL): string
+    {
+
+        // check presence of key
+        $trans = $this->translation[$context][$message] ?? false;
+
+        // if not present return string
+        if (false === $trans) {
+            return $message;
+        }
+
+        // if present but empty raise exception
+        if (empty($trans)) {
+            throw new Exception('Translation is present but is empty');
```

Comment:
I'd suggest returning a default string instead of an exception here, possibly just a static one eg. "No translation found", or the original text with that string added to it.

Reply:
is only an Exception that raise only for integrity when a translation was defined in a bad way.
Translation will be defined like this
```php
$translation['domain'] = [
    'stringA' => [
        0 => 'no string',
        1 => 'string',
        2 => 'strings',
    ]
];
```
It will raise only if defined like this :
```php
$translation['domain'] = [
    'stringA' => []
];
```

Reply:
Still feels wrong for me to print out exceptions in a translation class, since it can break otherwise valid code for a reason that's only partially worth it.
I'd suggest creating an error message instead and attaching it to the front of the translation (thus somewhat breaking it), but making it still useable in the worst-case scenario.

Reply:
I vote to define this through a parameter.

Reply:
like we do in App, is the best option, but in case of not raising exception, even if we return string untranslated or empty, we leave the user with no clue of where can be the error.

add flag $raise_bad_format_exception = boolean;
what returns? and empty string or the untranslated string?

Reply:
I’d log that and then add all translations automatically

Reply:
I'd suggest retuning the untranslated string.

Reply:
add flag : https://github.com/abbadon1334/core/blob/dc07b17f949fdea4f42c7b2b56351b5dd15736e1/src/Translator.php#L35